src/onboarding2/Create/Create.js

```diff
@@ -135,52 +144,146 @@ function useConfigureState({ onScreenUpdate }) {
+          }
+          return null
+        })
+        .catch(() => {
+          // Continuously re-request until this component gets unmounted or the template changes
+          if (!cancelled) {
+            fetchArtifact()
+          }
+        })
+    }
+
+    fetchArtifact()
+
+    return () => {
+      cancelled = true
+    }
+  }, [templateRepoAddress])
+
+  return {
+    fetchingTemplateInformation,
+    templateAbi,
+    templateAddress,
+  }
+}
+
+function useDeploymentState(
+  walletWeb3,
+  account,
+  status,
+  template,
+  templateData,
+  templateAbi,
+  templateAddress,
+  applyEstimateGas
+) {
+  const [transactionProgress, setTransactionProgress] = useState({
+    signing: 0,
+    error: -1,
+  })
 
   const deployTransactions = useMemo(
     () =>
-      status === STATUS_DEPLOYMENT
-        ? template.prepareTransactions(function createTransaction(name, args) {
-            // TODO
-            return [name, args]
-          }, templateData)
+      status === STATUS_DEPLOYMENT && templateAbi && templateAddress
+        ? template.prepareTransactions(
+            prepareTransactionCreatorFromAbi(templateAbi, templateAddress),
+            templateData
+          )
         : null,
-    [template, status, templateData]
+    [status, templateAbi, templateAddress, template, templateData]
   )
 
   // Call tx functions in the template, one after another.
   useEffect(() => {
-    setSignedTransactions(0)
+    setTransactionProgress({ signed: 0, errored: -1 })
 
     if (!deployTransactions) {
       return
     }
 
+    let cancelled = false
     const createTransactions = async () => {
-      for (const deployTransaction of deployTransactions) {
-        await new Promise(resolve =>
-          setTimeout(() => {
-            log('Transaction:', deployTransaction.transaction)
-            resolve()
-          }, 3000)
-        )
-        setSignedTransactions(signed => signed + 1)
-      }
+      // Only process the next transaction after the previous one was successfully mined
+      deployTransactions.reduce(async (deployPromise, { transaction }) => {
+        // Wait for the previous promise; if component has unmounted, don't progress any further
+        await deployPromise
+
+        transaction = {
+          ...transaction,
+          from: account,
+        }
+        try {
+          transaction = await applyEstimateGas(transaction)
+        } catch (_) {}
+
+        if (!cancelled) {
+          try {
+            await walletWeb3.eth.sendTransaction(transaction)
+
+            if (!cancelled) {
+              setTransactionProgress(({ signed, errored }) => ({
+                signed: signed + 1,
+                errored,
+              }))
+            }
+          } catch (err) {
+            log('Failed onboarding transaction', err)
+            if (!cancelled) {
+              setTransactionProgress(({ signed, errored }) => ({
+                errored: errored + 1,
```

Comment:
This will happen either due to the user cancelling the transaction, or to the transaction itself failing. We should handle this in `DeploymentSteps` and let the user go back to review (probably easiest?) or resend transactions.

Reply:
Yes I think it could be nice to have the two options? Resend a failed transaction, or come back to the review screen (or to the first configuration screen directly?). /cc @dizzypaty